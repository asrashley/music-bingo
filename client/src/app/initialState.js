import { initialState as cards } from '../cards/cardsSlice';
import { initialState as games } from '../games/gamesSlice';
import { initialState as tickets } from '../tickets/ticketsSlice';
import { initialState as user } from '../user/userSlice';

export const initialState = Object.freeze({
    cards,
    games,
    tickets,
    user,
});
