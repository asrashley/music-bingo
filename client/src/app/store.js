import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
