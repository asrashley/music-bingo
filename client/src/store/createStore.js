import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';

import adminReducer from '../admin/adminSlice';
import directoriesReducer from '../directories/directoriesSlice';
import gamesReducer from '../games/gamesSlice';
import messagesReducer from '../messages/messagesSlice';
import settingsReducer from '../settings/settingsSlice';
import systemReducer from '../system/systemSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import usersMiddleware from '../user/userMiddleware';
import messagesMiddleware from '../messages/messagesMiddleware';

import { history } from './history';

export function createStore(preloadedState, logging) {
  const middleware = [...getDefaultMiddleware(), usersMiddleware, messagesMiddleware];
  if (logging === undefined) {
    logging = (process.env.NODE_ENV === 'development');
  }
  if (logging) {
    middleware.push(logger);
  }
  return configureStore({
    middleware,
    reducer: {
      admin: adminReducer,
      directories: directoriesReducer,
      games: gamesReducer,
      messages: messagesReducer,
      settings: settingsReducer,
      system: systemReducer,
      tickets: ticketsReducer,
      router: connectRouter(history),
      user: userReducer,
    },
    preloadedState
  });
}
