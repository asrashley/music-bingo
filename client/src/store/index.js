export { createStore } from './createStore';
export { initialState } from './initialState';
export { history } from './history';
