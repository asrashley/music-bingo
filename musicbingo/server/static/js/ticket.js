$(document).ready(function() {
    "use strict";
    function load_state(index, ticket) {
        var game_id = $(ticket).data("game-id");
        var ticket_number = $(ticket).data("ticket-number");
        var key = "ticket-" + game_id + "-" + ticket_number;
        var data = localStorage.get(key);
        if (data !== null) {
            data = JSON.parse(data);
        }
    }

    $('.bingo-cell').on('click', function(ev) {
        console.log('click');
        $(ev.target).parentsUntil("tr").toggleClass('ticked');
    });
    $('.bingo-ticket').each(load_state);
});
