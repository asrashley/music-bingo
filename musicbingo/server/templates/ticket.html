<div class="view-ticket">
  <div class="download">
    <a href="{{url_for('download_ticket', game_pk=game.pk, ticket_pk=ticket.pk)}}"
       class="btn btn-primary btn-lg">Download ticket {{ticket.number}}</a>
  </div>
  <table class="bingo-ticket table table-striped table-bordered"
         data-game-id="{{game.id}}" data-ticket-number="{{ticket.number}}">
    <thead>
      <tr>
        <th colspan="5" class="logo">
        </th>
      </tr>
    </thead>
    <tbody>
      {% for row in ticket.rows %}
      <tr>
        {% for cell in row %}
        <td class="bingo-cell"
            data-row="{{cell.row}}" data-column="{{cell.column}}"
            style="background-color: {{cell.background.css()}}"
            >
          <div class="bingo-cell-wrap">
            <p class="title">{{cell.title}}</p>
            <p class="artist">{{cell.artist}}</p>
          </div>
        </td>
        {% endfor %}
      </tr>
      {% endfor %}
    </tbody>
    <tfoot>
      <tr>
        <th colspan="3" class="title">
          {{game.title}}
        </th>
        <th colspan="2" class="number">
          Game {{game.id}} / Ticket {{ticket.number}}
        </th>
      </tr>
    </tfoot>
  </table>
</div>
