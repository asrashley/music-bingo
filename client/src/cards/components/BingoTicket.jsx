import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { saveAs } from 'file-saver';

import { fetchCardIfNeeded,  setChecked } from '../../cards/cardsSlice';
import { getCard } from '../../cards/cardsSelectors';
import { initialState } from '../../app/initialState';
import { api } from '../../endpoints';

const BingoCell = ({ cell, onClick, options }) => {
  const tdStyle = {
    width: `${100 / options.columns}%`,
  };
  if (cell.background) {
    tdStyle.backgroundColor = cell.background;
  }
  let className = "bingo-cell";
  if (cell.checked === true) {
    className += " ticked";
  }
  return (
    <td className={className} onClick={(ev) => onClick(cell)}
      onTouchStart={(ev) => onClick(cell)}
      data-row={cell.row} data-column={cell.column}
      style={tdStyle}
    >
      <div className="bingo-cell-wrap">
        <p className="title">{cell.title}</p>
        <p className="artist">{cell.artist}</p>
      </div>
    </td>
  );
};

const TableRow = ({ row, onClick, options }) => {
  return (
    < tr >
      {row.map((cell, idx) => <BingoCell key={idx} cell={cell} options={options} onClick={onClick} />)}
    </tr>
  );
};

const CardError = ({ error }) => {
  return (
    <div class="card-error">{error}</div>
  );
};

class BingoTicket extends React.Component {
  static propTypes = {
    card: PropTypes.object.isRequired,
    game: PropTypes.object.isRequired,
    ticket: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired,
  };

  state = {
    error: null,
  };

  componentDidMount() {
    const { dispatch, game, ticket } = this.props;
    dispatch(fetchCardIfNeeded(game.pk, ticket.pk));
  }

  componentDidUpdate(prevProps, prevState) {
    const { dispatch, game, ticket } = this.props;
    if (prevProps.game.pk !== game.pk ||
      prevProps.ticket.pk !== ticket.pk) {
      dispatch(fetchCardIfNeeded(game.pk, ticket.pk));
    }
  }

  onClickCell = (cell) => {
    const { game, ticket, dispatch, user } = this.props;
    const { row, column } = cell;
    const number = row * user.options.columns + column;

    dispatch(setChecked({
      gamePk: game.pk,
      ticketPk: ticket.pk,
      row,
      column,
      number,
      checked: !cell.checked
    }));
  };

  downloadPDF = (ev) => {
    const { game, dispatch, ticket } = this.props;
    return dispatch(api.downloadCard({ gamePk: game.pk, ticketPk: ticket.pk }))
      .then((response) => {
        const filename = `Game ${game.id} - Ticket ${ticket.number}.pdf`;
        return response.payload.blob().then(blob => {
          saveAs(blob, filename);
          return filename;
        });
      })
      .catch(err => this.setState({ error: `${err}` }));
  };


  render() {
    const { game, ticket, card, user } = this.props;
    const { error } = this.state;
    /* <a href={getDownloadCardURL(game.pk, ticket.pk)}
      download={`Game ${game.id} - Ticket ${ticket.number}.pdf`}
      rel="_blank" type="application/pdf"
      className="btn btn-primary btn-lg">direct Download ticket {ticket.number}</a> */

    return (
      <div className="view-ticket">
        {error && <div className="alert alert-warning" role="alert"><span className="error-message">{error}</span></div>}
        <div className="download">
          <button className="btn btn-primary btn-lg" onClick={this.downloadPDF} >
            Download ticket {ticket.number}</button>
        </div>
        <table className="bingo-ticket table table-striped table-bordered"
          data-game-id={game.id} data-ticket-number={ticket.number}>
          <thead>
            <tr>
              <th colSpan={user.options.columns} className="logo">
              </th>
            </tr>
          </thead>
          <tbody>
            {card.rows.map((row, idx) => <TableRow key={idx} row={row} options={user.options} onClick={this.onClickCell} />)}
          </tbody>
          <tfoot>
            <tr>
              <th colSpan={user.options.columns - 2} className="title">
                {game.title}
              </th>
              <th colSpan="2" className="number">
                Game {game.id} / Ticket {ticket.number}
              </th>
            </tr>
          </tfoot>
        </table>
        {card.error && <CardError error={card.error} />}
      </div>
    );
  }
}

/*
const mapStateToProps = (state, ownProps) => {
  state = state || initialState;
  const { game, ticket } = ownProps;
  const { user } = state;
  return {
    user,
    game,
    ticket,
    card: getCard(state, ownProps),
  };
};

BingoTicket = connect(mapStateToProps)(BingoTicket);
*/

export {
  BingoTicket,
};
