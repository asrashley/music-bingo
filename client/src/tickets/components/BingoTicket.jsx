import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { saveAs } from 'file-saver';

import { fetchCardIfNeeded,  setChecked } from '../../cards/cardsSlice';
import { getCard } from '../../cards/cardsSelectors';
import { initialState } from '../../app/initialState';
import { getDownloadCardURL } from '../../endpoints';

const BingoCell = ({ cell, onClick, options }) => {
  const tdStyle = {
    width: `${100 / options.columns}%`,
  };
  if (cell.background) {
    tdStyle.backgroundColor = cell.background;
  }
  let className = "bingo-cell";
  if (cell.checked === true) {
    className += " ticked";
  }
  return (
    <td className={className} onClick={(ev) => onClick(cell)}
      onTouchStart={(ev) => onClick(cell)}
      data-row={cell.row} data-column={cell.column}
      style={tdStyle}
    >
      <div className="bingo-cell-wrap">
        <p className="title">{cell.title}</p>
        <p className="artist">{cell.artist}</p>
      </div>
    </td>
  );
};

const TableRow = ({ row, onClick, options }) => {
  return (
    < tr >
      {row.map((cell, idx) => <BingoCell key={idx} cell={cell} options={options} onClick={onClick} />)}
    </tr>
  );
};

const CardError = ({ error }) => {
  return (
    <div class="card-error">{error}</div>
  );
};

class BingoTicket extends React.Component {
  static propTypes = {
    card: PropTypes.object.isRequired,
    game: PropTypes.object.isRequired,
    ticket: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired,
  };

  componentDidMount() {
    const { dispatch, game, ticket } = this.props;
    dispatch(fetchCardIfNeeded(game.pk, ticket.pk));
  }

  componentDidUpdate(prevProps, prevState) {
    const { dispatch, game, ticket } = this.props;
    if (prevProps.game.pk !== game.pk ||
      prevProps.ticket.pk !== ticket.pk) {
      dispatch(fetchCardIfNeeded(game.pk, ticket.pk));
    }
  }

  onClickCell = (cell) => {
    const { game, ticket, dispatch } = this.props;
    const { row, column } = cell;

    dispatch(setChecked({
      gamePk: game.pk,
      ticketPk: ticket.pk,
      row,
      column,
      checked: !cell.checked
    }));
  };

  downloadPDF = (ev) => {
    const { game, ticket } = this.props;
    return fetch(getDownloadCardURL(game.pk, ticket.pk), {
      credentials: 'same-origin',
      header: {
        Accept: "application/pdf",
      }
    })
      .then((response) => {
        if (!response.ok) {
          return Promise.reject({ error: `${response.status}: ${response.statusText}` });
        }
        const filename = `Game ${game.id} - Ticket ${ticket.number}.pdf`;
        return response.blob().then(blob => {
          saveAs(blob, filename);
          return filename;
        });
      });
  };


  render() {
    const { game, ticket, card, user } = this.props;
    /* <a href={getDownloadCardURL(game.pk, ticket.pk)}
      download={`Game ${game.id} - Ticket ${ticket.number}.pdf`}
      rel="_blank" type="application/pdf"
      className="btn btn-primary btn-lg">direct Download ticket {ticket.number}</a> */

    return (
      <div className="view-ticket">
        <div className="download">
          <button className="btn btn-primary btn-lg" onClick={this.downloadPDF} >
            Download ticket {ticket.number}</button>
        </div>
        <table className="bingo-ticket table table-striped table-bordered"
          data-game-id={game.id} data-ticket-number={ticket.number}>
          <thead>
            <tr>
              <th colSpan={user.options.columns} className="logo">
              </th>
            </tr>
          </thead>
          <tbody>
            {card.rows.map((row, idx) => <TableRow key={idx} row={row} options={user.options} onClick={this.onClickCell} />)}
          </tbody>
          <tfoot>
            <tr>
              <th colSpan={user.options.columns - 2} className="title">
                {game.title}
              </th>
              <th colSpan="2" className="number">
                Game {game.id} / Ticket {ticket.number}
              </th>
            </tr>
          </tfoot>
        </table>
        {card.error && <CardError error={card.error} />}
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  state = state || initialState;
  const { game, ticket } = ownProps;
  return {
    game,
    ticket,
    card: getCard(state, ownProps),
  };
};

BingoTicket = connect(mapStateToProps)(BingoTicket);

export {
  BingoTicket,
};