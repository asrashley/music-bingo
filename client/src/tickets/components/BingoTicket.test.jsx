import React from 'react';
import { fireEvent } from '@testing-library/react';

import { renderWithProviders } from '../../../tests';
import { createStore } from '../../store/createStore';
import { initialState } from '../../store/initialState';
import { BingoTicket } from './BingoTicket';

import ticket from '../../../tests/fixtures/ticket.json';

describe('BingoTicket component', () => {
  const game = {
    "pk": 159,
    "id": "18-04-22-2",
    "title": "Various Artists",
    "slug": "various-artists",
    "start": "2018-04-22T15:43:56Z",
    "end": "2018-04-23T15:43:56Z",
    "options": {
      "colour_scheme": "blue",
      "number_of_cards": 36,
      "include_artist": true,
      "page_size": "A4",
      "columns": 5,
      "rows": 3,
      "checkbox": false,
      "cards_per_page": 4,
      "backgrounds": ["#daedff", "#f0f8ff", "#daedff", "#f0f8ff", "#daedff", "#f0f8ff", "#daedff", "#f0f8ff", "#daedff", "#f0f8ff", "#daedff", "#f0f8ff", "#daedff", "#f0f8ff", "#daedff"]
    },
    "userCount": 0
  };

  it('renders a ticket that is available', async () => {
    const props = {
      game,
      ticket,
      setChecked: vi.fn(),
      download: false,
    };
    props.ticket.game = game.pk;
    const store = createStore(initialState);
    props.dispatch = store.dispatch;
    const { getByText, asFragment } = renderWithProviders(<BingoTicket {...props} />, { store });
    props.ticket.rows.forEach((row) => {
      row.forEach(cell => {
        getByText(cell.title);
      });
    });
    const cell = getByText(props.ticket.rows[0][0].title);
    fireEvent.click(cell);
    expect(props.setChecked).toHaveBeenCalledTimes(1);
    expect(asFragment()).toMatchSnapshot();
  });
});
