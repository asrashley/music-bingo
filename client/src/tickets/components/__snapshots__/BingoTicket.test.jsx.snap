// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BingoTicket component renders a ticket that is available 1`] = `
<DocumentFragment>
  <div
    class="view-ticket"
  >
    <table
      class="bingo-ticket table"
      data-game-id="18-04-22-2"
      data-ticket-number="2"
    >
      <thead>
        <tr>
          <th
            class="logo"
            colspan="5"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Good Golly Miss Molly
              </p>
              <p
                class="artist"
              >
                Little Richard
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Hey! Baby
              </p>
              <p
                class="artist"
              >
                Bruce Channel
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Diana
              </p>
              <p
                class="artist"
              >
                Paul Anka
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Ain't That A Shame
              </p>
              <p
                class="artist"
              >
                Fats Domino
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Lipstick On Your Collar
              </p>
              <p
                class="artist"
              >
                Connie Francis
              </p>
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Only You
              </p>
              <p
                class="artist"
              >
                The Platters
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                As I Love You
              </p>
              <p
                class="artist"
              >
                Shirley Bassey
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Colette
              </p>
              <p
                class="artist"
              >
                Billy Fury
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Whole Lotta Woman
              </p>
              <p
                class="artist"
              >
                Marvin Rainwater
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Whole Lotta Shakin' Goin' On
              </p>
              <p
                class="artist"
              >
                Carl Perkins
              </p>
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Bird Dog
              </p>
              <p
                class="artist"
              >
                The Everly Brothers
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Stupid Cupid
              </p>
              <p
                class="artist"
              >
                Connie Francis
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                Stranger In Paradise
              </p>
              <p
                class="artist"
              >
                Tony Bennett
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                At The Hop
              </p>
              <p
                class="artist"
              >
                Danny & The Juniors
              </p>
            </div>
          </td>
          <td
            class="bingo-cell  with-artist"
            style="width: 20%;"
          >
            <div
              class="bingo-cell-wrap"
            >
              <p
                class="title"
              >
                C'mon Everybody
              </p>
              <p
                class="artist"
              >
                Eddie Cochran
              </p>
            </div>
          </td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <th
            class="footer-title"
            colspan="3"
          >
            Various Artists
          </th>
          <th
            class="footer-number"
            colspan="2"
          >
            Game 18-04-22-2 / Ticket 2
          </th>
        </tr>
      </tfoot>
    </table>
  </div>
</DocumentFragment>
`;
